src/DiagPDF.cpp

```diff
@@ -0,0 +1,367 @@
+{
+	// TODO(bwibking): skip fine-covered in search
+	amrex::Real mmin{AMREX_REAL_MAX};
+	for (const auto *st : a_state) {
+		mmin = std::min(mmin, st->min(comp, 0, true));
+	}
+
+	amrex::ParallelDescriptor::ReduceRealMin(mmin);
+	return mmin;
+}
+
+auto DiagPDF::MFVecMax(const amrex::Vector<const amrex::MultiFab *> &a_state, int comp) -> amrex::Real
+{
+	// TODO(bwibking): skip fine-covered in search
+	amrex::Real mmax{AMREX_REAL_LOWEST};
+	for (const auto *st : a_state) {
+		mmax = std::max(mmax, st->max(comp, 0, true));
+	}
+
+	amrex::ParallelDescriptor::ReduceRealMax(mmax);
+	return mmax;
+}
+
+void DiagPDF::writePDFToFile(int a_nstep, const amrex::Real &a_time, const amrex::Vector<amrex::Real> &a_pdf, const amrex::Real &a_sum)
+{
+	std::string diagfile;
+	if (m_interval > 0) {
+		diagfile = amrex::Concatenate(m_diagfile, a_nstep, 6);
+	}
+	if (m_per > 0.0) {
+		diagfile = m_diagfile + std::to_string(a_time);
+	}
+	diagfile = diagfile + ".dat";
+
+	if (amrex::ParallelDescriptor::IOProcessor()) {
+		std::ofstream pdfFile;
+		pdfFile.open(diagfile.c_str(), std::ios::out);
+		const int prec = 17;
+		const int width = 25;
+
+		const int nvars = static_cast<int>(m_varNames.size());
+		amrex::Vector<amrex::Vector<int>> widths(nvars);
+
+		for (int n = 0; n < nvars; ++n) {
+			widths[n].resize(3, width);
+			widths[n][0] = std::max(width, static_cast<int>(m_varNames[n].length()) + 5);
+			widths[n][1] = std::max(width, static_cast<int>(m_varNames[n].length()) + 5);
+			widths[n][2] = std::max(width, static_cast<int>(m_varNames[n].length()) + 5);
+
+			pdfFile << std::setw(widths[n][0]) << m_varNames[n] << "_idx"
+				<< " " << std::setw(widths[n][0]) << m_varNames[n] << "_min"
+				<< " " << std::setw(widths[n][1]) << m_varNames[n] << "_max";
+		}
+		pdfFile << " " << std::setw(width) << "PDF"
+			<< "\n";
+
+		std::vector<amrex::Real> transformed_range(nvars);
+		std::vector<amrex::Real> transformed_binWidth(nvars);
+		std::vector<amrex::Real> transformed_lowBnd(nvars);
+		for (int n = 0; n < nvars; ++n) {
+			transformed_range[n] =
+			    (m_useLogSpacedBins[n] != 0) ? (std::log10(m_highBnd[n]) - std::log10(m_lowBnd[n])) : (m_highBnd[n] - m_lowBnd[n]);
+			transformed_binWidth[n] = transformed_range[n] / m_nBins[n];
+			transformed_lowBnd[n] = (m_useLogSpacedBins[n] != 0) ? std::log10(m_lowBnd[n]) : m_lowBnd[n];
+		}
+
+		for (int linidx{0}; linidx < a_pdf.size(); ++linidx) {
+			std::vector<int> const idxVec = getIdxVec(linidx, m_nBins);
+			std::vector<amrex::Real> bin_min(nvars);
+			std::vector<amrex::Real> bin_max(nvars);
+			amrex::Real binvol = 1;
+
+			// calculate bin edges, bin volume
+			for (int n = 0; n < nvars; ++n) {
+				int const i = idxVec[n];
+				amrex::Real const transformed_bin_left = transformed_lowBnd[n] + static_cast<amrex::Real>(i) * transformed_binWidth[n];
+				amrex::Real const transformed_bin_right = transformed_bin_left + transformed_binWidth[n];
+				amrex::Real bin_left{NAN};
+				amrex::Real bin_right{NAN};
+
```

Comment:
There is a subtle question of how to normalize when your bins are logarithmically spaced, and your choice here is probably not the one I would make. If you have some variable x by which you're binning, and your bins are spaced logarithmically, then you have a choice about whether to define the width of a bin whose edges are x_1 and x_2 as log(x_2) - log(x_1) = log(x_2 / x_1), or to define it as just x_2 - x_1. The first choice corresponds to computing the probability distribution dp / d log(x), while the second to computing dp/dx over non-uniform bins. Both are valid choices, but personally I think it would make more sense, if the user requests uniformly-spaced logarithmic bins, to compute dp / d log(x) in uniform bins rather than dp / dx in non-uniform bins. If you agree, then you should remove this if clause, and just take bin_left and bin_right to be equal to transformed_bin_left and transformed_bin_right regardless of whether the bins are logarithmic or not. If you do want to keep this code as is, you should just very clearly document what setting the bins to logarithmic means, i.e., it means that even though the bins are logarithmically spaced, the quantity returned is still dp / dx and not dp / d log(x).

Reply:
The problem is that for multidimensional histograms, it's possible to have some bins that are log-spaced and others that are not, and we don't have a good way to indicate this in a CSV file.

In particular, there is no standard way to included parsable metadata in a csv file, so users are likely to ignore any comments at the top of the file that might attempt to explain this and instead make a severe units error when this scenario presents itself.

We could output YAML format files instead, and add metadata that specifies what the precise units are. This should solve the problem, since it will require writing significantly more code on the part of the user to parse the file, and they are more likely to have to carefully examine the format in order to get the histogram read back into memory. On the other hand, that is significantly more work.

Reply:
Actually, we could output the probability mass function instead, and let the user convert it to a density function with whatever units are preferred.

I am leaning toward this solution. Does this seem workable to you?

Reply:
See comment below for my proposed solution.